File: src/csharp/Foundgine.Core/Semantic/SemanticEntity.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic;

/// <summary>
/// The protocol-independent semantic description of an entity.
/// </summary>
public sealed record SemanticEntity(
    EntityId Id,
    string Name,
    SemanticFieldIdentity Identity,
    IReadOnlyList<SemanticField> Fields,
    IReadOnlyList<SemanticRelationship> Relationships,
    IReadOnlyList<SemanticAlias>? Aliases = null)
{
    public IReadOnlyList<SemanticAlias> EffectiveAliases => Aliases ?? [];

    /// <summary>CLR model type represented by this semantic entity when known.</summary>
    public Type? ModelType { get; init; }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticEntity.java
package com.foundgine.core.semantic;

import com.foundgine.core.abstractions.EntityId;
import java.util.*;

public record SemanticEntity(EntityId id, String name, SemanticFieldIdentity identity, List<SemanticField> fields,
		List<SemanticRelationship> relationships, List<SemanticAlias> aliases, Class<?> modelType) {
	public SemanticEntity {
		Objects.requireNonNull(id);
		Objects.requireNonNull(name);
		Objects.requireNonNull(identity);
		fields = List.copyOf(fields);
		relationships = List.copyOf(relationships);
		aliases = aliases == null ? List.of() : List.copyOf(aliases);
	}

	public SemanticEntity(EntityId id, String name, SemanticFieldIdentity identity, List<SemanticField> fields,
			List<SemanticRelationship> relationships) {
		this(id, name, identity, fields, relationships, List.of(), null);
	}

	public List<SemanticAlias> effectiveAliases() {
		return aliases;
	}
}
